### PR TITLE
Safe casting + math on commit epoch days calculation

### PR DESCRIPTION
**Pull request by mikiquantum**

### CODE REVIEW

File: pallets/anchors/src/common.rs [hunk 1]
```diff
@@ -21,14 +21,19 @@ use frame_support::storage::child::ChildInfo;
...
+	return match days_option {
+		None => None,
+		Some(v) => {
+			if remainder == 0 {
+				Some(v)
+			} else {
+				v.checked_add(1)
+			}
+		}
+	};
```
Comment (NunoAlexandre) [inline pallets/anchors/src/common.rs:36]: `til` - I think we can simplify this to:

```suggestion
	let days = u32::try_from(ts / MILLISECS_PER_DAY).ok()?;
  
  if remainder == 0 {
	  Some(days)
	 } else {
    days.checked_add(1)
   }
```

File: pallets/anchors/src/common.rs [hunk 2]
```diff
@@ -46,13 +51,16 @@ mod tests {
...
 		// 1971-01-01  00:00:01
-		assert_eq!(get_days_since_epoch(31536001000), 366);
+		assert_eq!(get_days_since_epoch(31536001000), Some(366));
 
 		// 1970-12-31  11:59:59
-		assert_eq!(get_days_since_epoch(31449600000), 364);
+		assert_eq!(get_days_since_epoch(31449600000), Some(364));
+
+		// Overflow Test with MAX u32 (after division)
+		assert_eq!(get_days_since_epoch(371085174374358017), None);
```
Comment (NunoAlexandre) [inline pallets/anchors/src/common.rs:63]:
```suggestion
		assert_eq!(get_days_since_epoch(u32::MAX), None);
```

File: pallets/anchors/src/tests.rs [hunk 3]
```diff
@@ -295,6 +295,22 @@ fn basic_commit() {
...
 		);
+
+		// commit anchor triggering days overflow
+		assert_noop!(
+			Anchors::commit(
+				Origin::signed(1),
+				pre_image2,
+				doc_root,
+				<Test as frame_system::Config>::Hashing::hash_of(&0),
+				371085174374358017 // triggers overflow
```
Comment (NunoAlexandre) [inline pallets/anchors/src/tests.rs:306]:
```suggestion
				u32::MAX // triggers overflow
```

File: pallets/anchors/src/tests.rs [hunk 4]
```diff
@@ -295,6 +295,22 @@ fn basic_commit() {
...
+			),
+			DispatchError::Module(ModuleError {
```
Comment (NunoAlexandre) [inline pallets/anchors/src/tests.rs:308]: Could we use the direct error type as suggested by https://github.com/centrifuge/centrifuge-chain/issues/919?

File: pallets/anchors/src/common.rs [hunk 5]
```diff
@@ -21,14 +21,19 @@ use frame_support::storage::child::ChildInfo;
...
-pub fn get_days_since_epoch(ts: u64) -> u32 {
+pub fn get_days_since_epoch(ts: u64) -> Option<u32> {
 	let remainder = ts % MILLISECS_PER_DAY;
-	let days = (ts / MILLISECS_PER_DAY) as u32;
-	if remainder == 0 {
-		days
-	} else {
-		days + 1
-	}
+	let days_option = u32::try_from(ts / MILLISECS_PER_DAY).ok();
```
Comment (lemunozm) [inline pallets/anchors/src/common.rs:26]: Maybe a `?`at the end saves you from writing the next match.

File: pallets/anchors/src/lib.rs [hunk 6]
```diff
@@ -292,10 +293,10 @@ pub mod pallet {
 			}
 
 			// pay the state rent
-			let today_in_days_from_epoch =
-				TryInto::<u64>::try_into(<pallet_timestamp::Pallet<T>>::get())
-					.map(common::get_days_since_epoch)
-					.or(Err(Error::<T>::FailedToConvertEpochToDays))?;
+			let now_u64 = TryInto::<u64>::try_into(<pallet_timestamp::Pallet<T>>::get())
+				.or(Err(Error::<T>::FailedToConvertEpochToDays))?;
```
Comment (lemunozm) [inline pallets/anchors/src/lib.rs:297]: FailedToConvertEpochToDays and EvictionDateTooBig errors could mean the same?

File: pallets/anchors/src/common.rs (hunk 1 above)
Comment (mikiquantum) [reply pallets/anchors/src/common.rs:36]: Nice, thanks I didnt know that `?` returned an option in addition to a result 👍

File: pallets/anchors/src/common.rs (hunk 2 above)
Comment (mikiquantum) [reply pallets/anchors/src/common.rs:63]: u32::MAX is `4294967295 ` which is smaller than the needed one that would trigger the overflow. This number `371085174374358017` is the minimum u64 needed to just go above the u32::MAX one.

File: pallets/anchors/src/lib.rs (hunk 6 above)
Comment (mikiquantum) [reply pallets/anchors/src/lib.rs:297]: Yeah, I will reuse ArithmeticError::Overflow

File: pallets/anchors/src/tests.rs (hunk 3 above)
Comment (mikiquantum) [reply pallets/anchors/src/tests.rs:306]: Same as above

File: pallets/anchors/src/tests.rs [hunk 7]
```diff
@@ -295,6 +295,22 @@ fn basic_commit() {
...
+				index: 5,
+				error: [8, 0, 0, 0],
+				message: Some("EvictionDateTooBig"),
+			})
```
[the remark this answers is not in the document]
Comment (mikiquantum) [reply pallets/anchors/src/tests.rs:312]: Yep, thanks, now that I am at it, I will change all the usages in this pallet.

File: pallets/anchors/src/common.rs (hunk 1 above)
Comment (NunoAlexandre) [reply pallets/anchors/src/common.rs]: yeah I think it's a somewhat new feature I believe :)

**Review by NunoAlexandre — APPROVED** [review]
💯

**Pull request merged**